File: backend/src/routes/przeglady.js
'use strict';
const express = require('express');
const { body } = require('express-validator');
const db = require('../config/db');
const { authenticate } = require('../middleware/auth');
const { roleGuard } = require('../middleware/roleGuard');
const { validate } = require('../middleware/validate');
const { writeAudit } = require('../services/auditService');

const router = express.Router();

router.get('/', authenticate, roleGuard('IT_ADMIN', 'SUPERADMIN'), async (req, res) => {
  const { tenant_id, status } = req.query;
  let q = `SELECT pr.*, t.nazwa as tenant_nazwa,
                  u.imie || ' ' || u.nazwisko as inicjujacy_nazwa,
                  COUNT(pp.id) as liczba_pozycji,
                  COUNT(pp.id) FILTER (WHERE pp.decyzja='POZOSTAW') as pozostaw,
                  COUNT(pp.id) FILTER (WHERE pp.decyzja='COFNIJ') as cofnij,
                  COUNT(pp.id) FILTER (WHERE pp.decyzja='ZMODYFIKUJ') as zmodyfikuj
             FROM przeglady pr
             JOIN tenants t ON t.id = pr.tenant_id
             JOIN uzytkownicy u ON u.id = pr.inicjujacy_id
             LEFT JOIN pozycje_przegladu pp ON pp.przeglad_id = pr.id
            WHERE 1=1`;
  const params = [];
  if (tenant_id) { params.push(tenant_id); q += ` AND pr.tenant_id=$${params.length}`; }
  if (status) { params.push(status); q += ` AND pr.status=$${params.length}`; }
  q += ' GROUP BY pr.id, t.nazwa, u.imie, u.nazwisko ORDER BY pr.data_rozpoczecia DESC';
  const { rows } = await db.query(q, params);
  res.json(rows);
});

router.get('/:id', authenticate, roleGuard('IT_ADMIN', 'SUPERADMIN'), async (req, res) => {
  const { rows } = await db.query(
    `SELECT pr.*, t.nazwa as tenant_nazwa,
            u.imie || ' ' || u.nazwisko as inicjujacy_nazwa
       FROM przeglady pr
       JOIN tenants t ON t.id = pr.tenant_id
       JOIN uzytkownicy u ON u.id = pr.inicjujacy_id
      WHERE pr.id=$1`,
    [req.params.id]
  );
  if (!rows.length) return res.status(404).json({ error: 'Przegląd nie istnieje.' });

  const { rows: pozycje } = await db.query(
    `SELECT pp.*,
            u.imie || ' ' || u.nazwisko as pracownik_nazwa,
            s.nazwa as system_nazwa, z.nazwa as zakres_nazwa, z.uprzywilejowany,
            ud.imie || ' ' || ud.nazwisko as decydent_nazwa
       FROM pozycje_przegladu pp
       JOIN uprawnienia up ON up.id = pp.uprawnienie_id
       JOIN pracownicy p ON p.id = pp.pracownik_id
       JOIN uzytkownicy u ON u.id = p.uzytkownik_id
       JOIN systemy_it s ON s.id = up.system_id
       JOIN zakres_uprawnien z ON z.id = up.zakres_id
       LEFT JOIN uzytkownicy ud ON ud.id = pp.decydent_id
      WHERE pp.przeglad_id=$1 ORDER BY u.nazwisko, s.nazwa`,
    [req.params.id]
  );

  res.json({ ...rows[0], pozycje });
});

// Initiate a review
router.post('/', authenticate, roleGuard('IT_ADMIN', 'SUPERADMIN'),
  [
    body('tenant_id').isInt(),
    body('typ').isIn(['CYKLICZNY', 'DORAZNY', 'ODEJSCIE']),
  ],
  validate,
  async (req, res) => {
    const { tenant_id, typ, uwagi, pracownik_id, system_id, komorka_id } = req.body;

    const client = await db.getClient();
    try {
      await client.query('BEGIN');
      const { rows: prRows } = await client.query(
        `INSERT INTO przeglady (tenant_id, typ, inicjujacy_id, uwagi)
         VALUES ($1,$2,$3,$4) RETURNING *`,
        [tenant_id, typ, req.user.userId, uwagi || null]
      );
      const przeglad = prRows[0];

      // Collect permissions to review
      let q = `SELECT up.id, up.pracownik_id FROM uprawnienia up
                 JOIN pracownicy p ON p.id = up.pracownik_id
                WHERE up.aktywne=TRUE AND up.tenant_id=$1`;
      const params = [tenant_id];
      if (pracownik_id) { params.push(pracownik_id); q += ` AND up.pracownik_id=$${params.length}`; }
      if (system_id) { params.push(system_id); q += ` AND up.system_id=$${params.length}`; }
      if (komorka_id) { params.push(komorka_id); q += ` AND p.komorka_id=$${params.length}`; }

      const { rows: ups } = await client.query(q, params);
      for (const up of ups) {
        await client.query(
          `INSERT INTO pozycje_przegladu (przeglad_id, uprawnienie_id, pracownik_id)
           VALUES ($1,$2,$3)`,
          [przeglad.id, up.id, up.pracownik_id]
        );
      }

      await client.query('COMMIT');
      await writeAudit({
        tenantId: tenant_id, userId: req.user.userId, username: req.user.username, rola: req.user.rola,
        akcja: 'PRZEGLAD_INICJOWANY', tabelaDocelowa: 'przeglady', rekordId: przeglad.id,
        noweDane: { typ, liczba_pozycji: ups.length },
      });
      res.status(201).json({ ...przeglad, liczba_pozycji: ups.length });
    } catch (err) {
      await client.query('ROLLBACK');
      throw err;
    } finally {
      client.release();
    }
  }
);

// Record decision for a review item
router.patch('/pozycje/:id/decyzja', authenticate, roleGuard('IT_ADMIN', 'SUPERADMIN'),
  [body('decyzja').isIn(['POZOSTAW', 'COFNIJ', 'ZMODYFIKUJ']), body('uzasadnienie').notEmpty()],
  validate,
  async (req, res) => {
    const { decyzja, uzasadnienie } = req.body;
    const { rows } = await db.query(
      `UPDATE pozycje_przegladu SET decyzja=$1, uzasadnienie=$2,
         data_decyzji=NOW(), decydent_id=$3 WHERE id=$4 RETURNING *`,
      [decyzja, uzasadnienie, req.user.userId, req.params.id]
    );
    if (!rows.length) return res.status(404).json({ error: 'Pozycja nie istnieje.' });

    // Execute decision
    if (decyzja === 'COFNIJ') {
      await db.query(
        `UPDATE uprawnienia SET aktywne=FALSE, data_cofniecia=NOW(),
           cofniete_przez=$1, powod_cofniecia=$2
         WHERE id=$3`,
        [req.user.userId, uzasadnienie, rows[0].uprawnienie_id]
      );
      // Update permission review result
      await db.query(
        `UPDATE uprawnienia SET wynik_przegladu='COFNIJ', data_ostatniego_przegladu=CURRENT_DATE,
           wymaga_przegladu=FALSE WHERE id=$1`,
        [rows[0].uprawnienie_id]
      );
    } else {
      await db.query(
        `UPDATE uprawnienia SET wynik_przegladu=$1, data_ostatniego_przegladu=CURRENT_DATE,
           wymaga_przegladu=FALSE WHERE id=$2`,
        [decyzja, rows[0].uprawnienie_id]
      );
    }
    res.json(rows[0]);
  }
);

// Close review
router.post('/:id/zakoncz', authenticate, roleGuard('IT_ADMIN', 'SUPERADMIN'), async (req, res) => {
  const { rows } = await db.query(
    `UPDATE przeglady SET status='ZAKOŃCZONY', data_zakonczenia=NOW() WHERE id=$1 RETURNING *`,
    [req.params.id]
  );
  if (!rows.length) return res.status(404).json({ error: 'Przegląd nie istnieje.' });
  res.json(rows[0]);
});

// Cancel review
router.post('/:id/anuluj', authenticate, roleGuard('IT_ADMIN', 'SUPERADMIN'), async (req, res) => {
  const { rows } = await db.query(
    `UPDATE przeglady SET status='ANULOWANY', data_zakonczenia=NOW() WHERE id=$1 RETURNING *`,
    [req.params.id]
  );
  if (!rows.length) return res.status(404).json({ error: 'Przegląd nie istnieje.' });
  res.json(rows[0]);
});

module.exports = router;

File: backend-py/routes/przeglady.py
from __future__ import annotations
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel
from sqlalchemy import text
from sqlalchemy.orm import Session

from database import get_session
from dependencies import CurrentUser, require_roles, write_audit, get_client_ip

router = APIRouter()
IT = ("IT_ADMIN", "SUPERADMIN")


class PrzegladIn(BaseModel):
    tenant_id: int
    typ: str
    uwagi: Optional[str] = None
    pracownik_id: Optional[int] = None
    system_id: Optional[int] = None
    komorka_id: Optional[int] = None


class DecyzjaIn(BaseModel):
    decyzja: str
    uzasadnienie: str


@router.get("/")
async def list_przeglady(tenant_id: Optional[int] = None, status: Optional[str] = None,
                          session: Session = Depends(get_session),
                          current_user: CurrentUser = Depends(require_roles(*IT))):
    q = """SELECT pr.*, t.nazwa as tenant_nazwa, u.imie || ' ' || u.nazwisko as inicjujacy_nazwa,
                  COUNT(pp.id) as liczba_pozycji,
                  COUNT(CASE WHEN pp.decyzja='POZOSTAW' THEN 1 END) as pozostaw,
                  COUNT(CASE WHEN pp.decyzja='COFNIJ' THEN 1 END) as cofnij,
                  COUNT(CASE WHEN pp.decyzja='ZMODYFIKUJ' THEN 1 END) as zmodyfikuj
             FROM przeglady pr
             JOIN tenants t ON t.id=pr.tenant_id
             JOIN uzytkownicy u ON u.id=pr.inicjujacy_id
             LEFT JOIN pozycje_przegladu pp ON pp.przeglad_id=pr.id
            WHERE 1=1"""
    params: dict = {}
    if tenant_id:
        q += " AND pr.tenant_id=:tid"
        params["tid"] = tenant_id
    if status:
        q += " AND pr.status=:s"
        params["s"] = status
    q += " GROUP BY pr.id, t.nazwa, u.imie, u.nazwisko ORDER BY pr.data_rozpoczecia DESC"
    rows = session.execute(text(q), params).mappings().all()
    return [dict(r) for r in rows]


@router.get("/{przeglad_id}")
async def get_przeglad(przeglad_id: int, session: Session = Depends(get_session),
                        current_user: CurrentUser = Depends(require_roles(*IT))):
    row = session.execute(
        text("""SELECT pr.*, t.nazwa as tenant_nazwa, u.imie || ' ' || u.nazwisko as inicjujacy_nazwa
                  FROM przeglady pr
                  JOIN tenants t ON t.id=pr.tenant_id
                  JOIN uzytkownicy u ON u.id=pr.inicjujacy_id
                 WHERE pr.id=:id"""),
        {"id": przeglad_id},
    ).mappings().first()
    if not row:
        raise HTTPException(status_code=404, detail="Przegląd nie istnieje.")
    result = dict(row)

    pozycje = session.execute(
        text("""SELECT pp.*, u.imie || ' ' || u.nazwisko as pracownik_nazwa,
                       s.nazwa as system_nazwa, z.nazwa as zakres_nazwa, z.uprzywilejowany,
                       ud.imie || ' ' || ud.nazwisko as decydent_nazwa
                  FROM pozycje_przegladu pp
                  JOIN uprawnienia up ON up.id=pp.uprawnienie_id
                  JOIN pracownicy p ON p.id=pp.pracownik_id
                  JOIN uzytkownicy u ON u.id=p.uzytkownik_id
                  JOIN systemy_it s ON s.id=up.system_id
                  JOIN zakres_uprawnien z ON z.id=up.zakres_id
                  LEFT JOIN uzytkownicy ud ON ud.id=pp.decydent_id
                 WHERE pp.przeglad_id=:id ORDER BY u.nazwisko, s.nazwa"""),
        {"id": przeglad_id},
    ).mappings().all()
    result["pozycje"] = [dict(p) for p in pozycje]
    return result


@router.post("/", status_code=201)
async def create_przeglad(body: PrzegladIn, request: Request,
                           session: Session = Depends(get_session),
                           current_user: CurrentUser = Depends(require_roles(*IT))):
    from datetime import datetime
    now = datetime.utcnow()
    pr_row = session.execute(
        text("INSERT INTO przeglady (tenant_id,typ,inicjujacy_id,uwagi,data_rozpoczecia) VALUES (:tid,:t,:iid,:u,:dt) RETURNING *"),
        {"tid": body.tenant_id, "t": body.typ, "iid": current_user.userId, "u": body.uwagi, "dt": now},
    ).mappings().first()
    session.commit()
    przeglad = dict(pr_row)

    q = "SELECT up.id, up.pracownik_id FROM uprawnienia up JOIN pracownicy p ON p.id=up.pracownik_id WHERE up.aktywne=1 AND up.tenant_id=:tid"
    params: dict = {"tid": body.tenant_id}
    if body.pracownik_id:
        q += " AND up.pracownik_id=:pid"
        params["pid"] = body.pracownik_id
    if body.system_id:
        q += " AND up.system_id=:sid"
        params["sid"] = body.system_id
    if body.komorka_id:
        q += " AND p.komorka_id=:kid"
        params["kid"] = body.komorka_id

    ups = session.execute(text(q), params).mappings().all()
    for up in ups:
        session.execute(
            text("INSERT INTO pozycje_przegladu (przeglad_id,uprawnienie_id,pracownik_id) VALUES (:prid,:uid,:pid)"),
            {"prid": przeglad["id"], "uid": up["id"], "pid": up["pracownik_id"]},
        )
    session.commit()

    write_audit(session, tenant_id=body.tenant_id, user_id=current_user.userId,
                username=current_user.username, rola=current_user.rola,
                akcja="PRZEGLAD_INICJOWANY", tabela_docelowa="przeglady", rekord_id=przeglad["id"],
                nowe_dane={"typ": body.typ, "liczba_pozycji": len(ups)}, ip_adres=get_client_ip(request))
    przeglad["liczba_pozycji"] = len(ups)
    return przeglad


@router.patch("/pozycje/{pozycja_id}/decyzja")
async def decyzja(pozycja_id: int, body: DecyzjaIn, request: Request,
                   session: Session = Depends(get_session),
                   current_user: CurrentUser = Depends(require_roles(*IT))):
    from datetime import datetime
    row = session.execute(
        text("UPDATE pozycje_przegladu SET decyzja=:d, uzasadnienie=:u, data_decyzji=:dt, decydent_id=:did WHERE id=:id RETURNING *"),
        {"d": body.decyzja, "u": body.uzasadnienie, "dt": datetime.utcnow(), "did": current_user.userId, "id": pozycja_id},
    ).mappings().first()
    if not row:
        raise HTTPException(status_code=404, detail="Pozycja nie istnieje.")
    session.commit()
    result = dict(row)

    if body.decyzja == "COFNIJ":
        session.execute(
            text("UPDATE uprawnienia SET aktywne=0, data_cofniecia=:dt, cofniete_przez=:uid, powod_cofniecia=:u WHERE id=:id"),
            {"dt": datetime.utcnow(), "uid": current_user.userId, "u": body.uzasadnienie, "id": result["uprawnienie_id"]},
        )
        session.execute(
            text("UPDATE uprawnienia SET wynik_przegladu='COFNIJ', data_ostatniego_przegladu=:dt, wymaga_przegladu=0 WHERE id=:id"),
            {"dt": datetime.utcnow().date(), "id": result["uprawnienie_id"]},
        )
    else:
        session.execute(
            text("UPDATE uprawnienia SET wynik_przegladu=:d, data_ostatniego_przegladu=:dt, wymaga_przegladu=0 WHERE id=:id"),
            {"d": body.decyzja, "dt": datetime.utcnow().date(), "id": result["uprawnienie_id"]},
        )
    session.commit()
    return result


@router.post("/{przeglad_id}/zakoncz")
async def zakoncz(przeglad_id: int, session: Session = Depends(get_session),
                   current_user: CurrentUser = Depends(require_roles(*IT))):
    from datetime import datetime
    row = session.execute(
        text("UPDATE przeglady SET status='ZAKOŃCZONY', data_zakonczenia=:dt WHERE id=:id RETURNING *"),
        {"dt": datetime.utcnow(), "id": przeglad_id},
    ).mappings().first()
    session.commit()
    if not row:
        raise HTTPException(status_code=404, detail="Przegląd nie istnieje.")
    return dict(row)


@router.post("/{przeglad_id}/anuluj")
async def anuluj(przeglad_id: int, session: Session = Depends(get_session),
                  current_user: CurrentUser = Depends(require_roles(*IT))):
    from datetime import datetime
    row = session.execute(
        text("UPDATE przeglady SET status='ANULOWANY', data_zakonczenia=:dt WHERE id=:id RETURNING *"),
        {"dt": datetime.utcnow(), "id": przeglad_id},
    ).mappings().first()
    session.commit()
    if not row:
        raise HTTPException(status_code=404, detail="Przegląd nie istnieje.")
    return dict(row)
